subprojects/packagefiles/openssl/meson.build

```diff
@@ -0,0 +1,482 @@
+project(
+  'openssl',
+  'c',
+  version: '1.1.1k',
+  license : 'Apache-2.0',
+  meson_version: '>= 0.55',
+  default_options: [
+    'warning_level=1',
+    'optimization=3',
+  ],
+)
+
+include_directories = [
+  'include',
+  'crypto',
+  'crypto/modes',
+  'crypto/ec/curve448',
+  'crypto/ec/curve448/arch_32',
+]
+
+compiler = meson.get_compiler('c')
+
+dependencies = []
+defines = []
+c_args = []
+
+defines += [
+  # Compile out hardware engines.  Most are stubs that dynamically load
+  # the real driver but that poses a security liability when an attacker
+  # is able to create a malicious DLL in one of the default search paths.
+  'OPENSSL_NO_HW',
+]
+
+asm = 'no-asm'
+if get_option('openssl_asm')
+  gas = find_program(
+    'as',
+    required: false,
+  )
```

Comment:
Meson can check the version for you: `gas = find_program('as', required: false, version: '>=2.26')`

Reply:
I would even do:
```
asm_opt = get_option('openssl_asm')
gas = find_program('as', required: asm_opt, version: '>=2.26')
nasm = find_program('nasm', required: asm_opt, version: '>=2.11.8')
if gas.found() and nasm.found()
  ...
else
  ...
endif
```

And make `openssl_asm` a feature option that defaults to `auto`.

Reply:
This isn't what code does right now. There are 3 invariants:
1) No gas/nasm
2) New gas/nasm
3) Old gas/nasm

If we require specific version, we can't support `3)` here, which is why I ended up with this mess.

Reply:
You can check for it twice, if you don't find the new version with required: false you can check again with a lower version number.

Reply:
OK, I have refactored things, hopefully it is more readable now. But the approach suggested above wouldn't work since it needs either gas or nasm, it doesn't need both at the same time, so it doesn't make sense to just use `required` on both.

Reply:
Something like that:

```
asm_opt = get_option('openssl_asm')
gas = find_program('as', required: asm_opt)
nasm = find_program('nasm', required: asm_opt)
new_gas = find_program('as', required: false, version: '>=2.26')
new_nasm = find_program('nasm', required: false, version: '>=2.11.8')
if new_gas.found() and new_nasm.found()
  # new asm
elif gas.found() and nasm.found()
  # old asm
else
  # no asm
endif
```

Reply:
That snippet will fail if you have nasm installed, but not gas. Only one of them is sufficient, but effective `required: true` will cause unnecessary early error:
```
openssl| Program as found: NO

../subprojects/openssl-1.1.1k/meson.build:41:2: ERROR: Program 'as' not found
```

Reply:
But currently you do:

```meson

  new_gas_or_nasm = (
    find_program('as', version: '>=2.26').found() or
    find_program('nasm', version: '>=2.11.8').found()
  )
```

which is defaulting to `required: true`.

Also, where is the gas or nasm program actually invoked? It's processed by the c_COMPILER build rule.

Reply:
Something like if asm_opt.enabled() and not as.found and not Naam.found error() endif

Reply:
I made both non-optional.

It is not called explicitly from Meson, but I think it is used internally somewhere. On Linux it is pretty difficult to end up not having `as`, but have C compiler at the same time, but I have honestly no idea what it looks like on other platforms and this part comes from Node.js side almost 1 to 1. Then did have a few strange decisions there that I tweaked, but not this part.

Having both non-optional was indeed a mistake, just fixed it.

Reply:
I like how it's done now, closing this thread.